test/functional/wallet_conflicts.py

```diff
@@ -123,5 +129,148 @@ def run_test(self):
+        assert_equal(alice.getbalance(), 25)
+
+        tx2 = alice.sendrawtransaction(tx2)
+
+        # Check that the 0th unspent is now available because the transaction spending it has been replaced in the mempool
+        assert_equal(alice.listunspent()[0]["txid"], unspents[0]["txid"])
+        assert_equal(alice.getbalance(), 25)
+
+        assert_equal(alice.gettransaction(tx1)["mempoolconflicts"], [tx2])
+
+        self.log.info("Test scenario where a mempool conflict is removed")
+
+        tx3 = alice.sendrawtransaction(tx3)
+
+        assert_equal(alice.gettransaction(tx1)["mempoolconflicts"], [])
+
+        # now all of alice's outputs should be considered spent
+        assert_equal(alice.listunspent(), [])
+        assert_equal(alice.getbalance(), 0)
+
+        alice.sendrawtransaction(alice.gettransaction(tx1)['hex'])
+        self.generate(self.nodes[0], 3)
+        assert_equal(alice.getbalance(), 75)
+
+    def test_mempool_and_block_conflicts(self):
+        alice = self.nodes[0].get_wallet_rpc("alice")
+        bob = self.nodes[1].get_wallet_rpc("bob")
+
+        self.log.info("Test a scenario where a transaction has both a block conflict and a mempool conflict")
+        unspents = [{"txid" : element["txid"], "vout" : element["vout"]} for element in alice.listunspent()]
+
+        assert_equal(bob.getbalances()["mine"]["untrusted_pending"], 0)
+
+        self.disconnect_nodes(0, 1)
+
+        raw_tx = alice.createrawtransaction(inputs=[unspents[0]], outputs=[{bob.getnewaddress() : 24.99999}])
+        raw_tx1 = alice.signrawtransactionwithwallet(raw_tx)['hex']
+        tx1 = bob.sendrawtransaction(raw_tx1)
+
+        raw_tx = alice.createrawtransaction(inputs=[unspents[0], unspents[2]], outputs=[{alice.getnewaddress() : 49.999}])
+        tx1_conflict = alice.signrawtransactionwithwallet(raw_tx)['hex']
+
+        raw_tx = alice.createrawtransaction(inputs=[unspents[2]], outputs=[{alice.getnewaddress() : 24.99}])
+        tx1_conflict_conflict = alice.signrawtransactionwithwallet(raw_tx)['hex']
+
+        raw_tx = alice.createrawtransaction(inputs=[unspents[1]], outputs=[{bob.getnewaddress() : 24.9999}])
+        raw_tx2 = alice.signrawtransactionwithwallet(raw_tx)['hex']
+        tx2 = bob.sendrawtransaction(raw_tx2)
+
+        raw_tx = alice.createrawtransaction(inputs=[unspents[1]], outputs=[{alice.getnewaddress() : 24.9999}])
+        tx2_conflict = alice.signrawtransactionwithwallet(raw_tx)['hex']
+
+        bob_unspents = [{"txid" : element, "vout" : 0} for element in [tx1, tx2]]
+
+        assert_equal(bob.getbalances()["mine"]["untrusted_pending"], Decimal("49.99989000"))
+
+        raw_tx = bob.createrawtransaction(inputs=[bob_unspents[0], bob_unspents[1]], outputs=[{bob.getnewaddress() : 49.999}])
+        raw_tx3 = bob.signrawtransactionwithwallet(raw_tx)['hex']
+        tx3 = bob.sendrawtransaction(raw_tx3)
+
+        # alice has 0 txs, bob has 3
```

Comment:
In 003efbbe45079c4416810a025b2bc372559dff15 "test: Add tests for wallet mempool conflicts":
I’m having trouble keeping track of all the things going on here. Perhaps you could add in these comments which transactions each node now knows about

Reply:
I agree that it gets a bit confusing, so I've added a lot more comments.